doc/design/float16.md

```diff
@@ -0,0 +1,46 @@
+# Design Doc: float16
+
+## Why float16
+Half precision (float16) is a binary floating-point format that occupies 16 bits / 2 bytes in memory. float16 is half the size of traditional 32-bit single precision format (float) and has lower precision and smaller range. 
+
+When high precision computation is not required, using float16 data type could potentially 
+
+- reduce storage space, memory bandwidth, and power usages; 
+- increase the chance of data fitting into a smaller cache of lower latency; 
+- provide arithmetic speed up if supported by hardware. 
+
+A brief survey of float16 support on different hardwares can be found [here](https://github.com/PaddlePaddle/Paddle/issues/4853). A brief survey of existing float16 implementations can be found [here](https://github.com/Xreki/Xreki.github.io/blob/master/multi_data_types_in_dl_framework/ppt/float16_and_quantized_type.md). 
+
+There are various natively supported float16 implementations on different hardwares/linear algebra libraries including half on cuda, float16_t on ARM processor, and Eigen::half on Eigen.
```

Comment:
I think a more detailed description is needed here. Need to describe the support of float16 from the three levels of hardware, compiler, and library respectively.
- For nvcc compiler, after CUDA 7.5, the __half type is supported.
- For NVIDIA GPU, maybe after sm_6.0(or sm_5.3?)
- For gcc/clang? In the mobile, usually, clang compiler.
- For libraries. Currently, what libraries support float16 calculations?
This is important because of if we upgrade the environment, we need to know what the minimum environment to support is.

Reply:
Great point! A detailed description has been added. Thanks!